feat(treesitter)!: deprecate top level indexes to modules

lewis6991 opened the pull request:
• The following top level Treesitter functions have been moved:
  `vim.treesitter.inspect_language()`    -> `vim.treesitter.language.inspect()`
  `vim.treesitter.get_query_files()`     -> `vim.treesitter.query.get_files()`
  `vim.treesitter.set_query()`           -> `vim.treesitter.query.set()`
  `vim.treesitter.query.set_query()`     -> `vim.treesitter.query.set()`
  `vim.treesitter.get_query()`           -> `vim.treesitter.query.get()`
  `vim.treesitter.query.get_query()`     -> `vim.treesitter.query.get()`
  `vim.treesitter.parse_query()`         -> `vim.treesitter.query.parse()`
  `vim.treesitter.query.parse_query()`   -> `vim.treesitter.query.parse()`
  `vim.treesitter.add_predicate()`       -> `vim.treesitter.query.add_predicate()`
  `vim.treesitter.add_directive()`       -> `vim.treesitter.query.add_directive()`
  `vim.treesitter.list_predicates()`     -> `vim.treesitter.query.list_predicates()`
  `vim.treesitter.list_directives()`     -> `vim.treesitter.query.list_directives()`
  `vim.treesitter.query.get_range()`     -> `vim.treesitter.get_range()`
  `vim.treesitter.query.get_node_text()` -> `vim.treesitter.get_node_text()`

clason commented:
Can you add a "before -- after" to the PR description? It's not immediately obvious from the code change what this PR means to users.

lewis6991 commented:
Very draft, and might not go with this, but basically I don't like how `vim.treesitter.*` is an alias for `vim.treesitter.query.*` and `vim.treesitter.language.*`.

`query.lua` also has some functions that should be in `treesitter.lua`.

lewis6991 committed:
feat(treesitter)!: deprecate top level indexes to modules

The following top level Treesitter functions have been moved:
  - vim.treesitter.inspect_language() -> vim.treesitter.language.inspect()
  - vim.treesitter.get_query_files() -> vim.treesitter.query.get_files()
  - vim.treesitter.set_query() -> vim.treesitter.query.set()
  - vim.treesitter.query.set_query() -> vim.treesitter.query.set()
  - vim.treesitter.get_query() -> vim.treesitter.query.get()
  - vim.treesitter.query.get_query() -> vim.treesitter.query.get()
  - vim.treesitter.parse_query() -> vim.treesitter.query.parse()
  - vim.treesitter.query.parse_query() -> vim.treesitter.query.parse()
  - vim.treesitter.add_predicate() -> vim.treesitter.query.add_predicate()
  - vim.treesitter.add_directive() -> vim.treesitter.query.add_directive()
  - vim.treesitter.list_predicates() -> vim.treesitter.query.list_predicates()
  - vim.treesitter.list_directives() -> vim.treesitter.query.list_directives()
  - vim.treesitter.query.get_range() -> vim.treesitter.get_range()
  - vim.treesitter.query.get_node_text() -> vim.treesitter.get_node_text()

justinmk commented:
> basically I don't like how `vim.treesitter.*` is an alias for `vim.treesitter.query.*` and `vim.treesitter.language.*`.

As an interface, I see `vim.treesitter.query.*` as mostly a cost. Having the interface at one level feels more ergonomic, at least currently. Having to dig around in various submodules doesn't have much benefit. The fact that the impl lives in a submodule is an impl detail.

If we expect "query" to become a major topic with many functions, then yes it makes sense for the interface to live at `vim.treesitter.query.*`. But I don't think we expect that?

clason commented:
> If we expect "query" to become a major topic with many functions, then yes it makes sense for the interface to live at vim.treesitter.query.*. But I don't think we expect that? Especially if we have object methods: the interface is the object class, not the module.

It _is_ a major topic, as this is basically the interface to our "runtime" files, including customization (currently 10 exported functions). Similar arguments can be made about (some of) the other submodules.

I know we are of a different mind on this in general, but I believe the grouping _helps_  discoverability, especially if the documentation can be improved to keep the functions with the topic introduction. (This is one of the drawbacks of the current "throwing everything under `vim.treesitter` approach.)

In a nutshell, tree-sitter has a lot of moving parts, and pretending it does not by just having a single `vim.treesitter`makes things harder to understand. So I kinda disagree that these are _just_ implementation details.

lewis6991 commented:
I don't mind either way, what I do mind is that we have both: `vim.treesitter.get_query` and `vim.treesitter.query.get_query()`.

I'm happy with any solution that resolves that.

clason commented:
My main concern is with the documentation, where I would like to conserve/improve the grouping by topic. From the point of view of writing code, `vim.treesitter.get_query` is fine with me (and admittedly nicer to type and read than `vim.treesitter.query.get_query`.

And I still don't know which of these two your PR deprecates ;)

lewis6991 commented:
I've listed all the changes in news.

justinmk commented:
> I don't mind either way, what I do mind is that we have both: `vim.treesitter.get_query` and `vim.treesitter.query.get_query()`.

I didn't think we exposed `vim.treesitter.query.get_query()` -- it's not documented (thus private), and `vim.treesitter.qu<tab>` doesn't auto-complete.

> Note that applications are already doing `local query = vim.treesitter.query`.

Oh, that means we don't avoid churn by removing `vim.treesitter.query`.

I agree we shouldn't have two different interfaces, so no objection to deprecating the top-level interface. And we need to fix `vim.treesitter.qu<tab>` ?

lewis6991 commented:
I'm very driven by making sure our code works well with the LSP. If anything the top level redirection stuff makes it harder to annotate our code properly.

clason commented:
For the record, I'm fine with the `vim.treesitter.verb_foo()` -> `vim.treesitter.foo.verb()` changes.

Can we change the docgen to put the module function reference to the corresponding module section (where this is not already the case)?

lewis6991 commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -9,12 +7,9 @@ local parsers = setmetatable({}, { __mode = 'v' })
-
-setmetatable(M, {
+---@field query TSQueryModule
+---@field language TSLanguageModule
```
This fixes completion.

lewis6991 commented:
> For the record, I'm fine with the vim.treesitter.verb_foo() -> vim.treesitter.foo.verb() changes.

I think that's all done. AFAIK it just needed a minor change in `gen_vimdoc.py`.

lewis6991 commented:
Yeah, I'm sure the tests are very broken.

clason commented:
Tests look good now; it's just complaining about an invalid help tag (`vim.treesitter.add_predicate()`).

clason commented on runtime/doc/news.txt:
```diff
@@ -275,11 +275,28 @@ DEPRECATIONS                                                *news-deprecations*
+  `vim.treesitter.query.get_query()`     -> `vim.treesitter.query.get()`
+  `vim.treesitter.parse_query()`         -> `vim.treesitter.query.parse()`
+  `vim.treesitter.query.parse_query()`   -> `vim.treesitter.query.parse()`
+  `vim.treesitter.get_node_text()`       -> `vim.treesitter.query.get_node_text()`
```
so this

clason commented on runtime/doc/news.txt:
```diff
@@ -275,11 +275,28 @@ DEPRECATIONS                                                *news-deprecations*
+  `vim.treesitter.list_predicates()`     -> `vim.treesitter.query.list_predicates()`
+  `vim.treesitter.list_directives()`     -> `vim.treesitter.query.list_directives()`
+  `vim.treesitter.query.get_range()`     -> `vim.treesitter.get_range()`
+  `vim.treesitter.query.get_node_text()` -> `vim.treesitter.get_node_text()`
```
and this leave things unchanged?
